Convert client.List to use functional options

grantr opened the pull request:
Replaces:

`List(ctx, ListOptions, list)`

with:

`List(ctx, list, ...Option)`

This matches the functional options signature of `client.Delete` in #94.

A few examples (I tried formatting these as a table but it wasn't wide enough):

Select all:
```go
// current
List(ctx, nil, list)
// proposed
List(ctx, list)
```

Filter by namespace:
```go
// current
List(ctx, client.InNamespace(ns), list)
// proposed
List(ctx, list, client.InNamespace(ns))
```

Filter by namespace and labels:
```go
// current
List(ctx, client.InNamespace(ns).MatchingLabels(labels), list)
// proposed
List(ctx, list, client.InNamespace(ns), client.MatchingLabels(labels))
```

To assemble a `ListOptions` with the builder methods or some other strategy, use the `UseListOptions` functional option:

```go
lo := &client.ListOptions{}
client.List(ctx, list, client.UseListOptions(
  lo.InNamespace(ns).MatchingLabels(labels)
))
```

DirectXMan12 commented:
@pwittrock what's our policy on breaking backwards compat?  This ends up with a much more ergonomic interface, and while I'm generally loathe to break the compatibility, I think it ends up being much less verbose, and follows the "you don't have to care about what you're not using" principle better.

droot commented:
@grantr new interface is nice UX wise, but can't get break the backward compatibility.

@DirectXMan12 We haven't defined the policy yet and have not been versioning the lib yet. Are there any good library projects we can take inspiration from wrt versioning/backward-compatibility aspect ?

grantr commented:
think the backward compatibility policy dovetails with the dependency versioning story from #111. How much is controller-runtime a standalone library versus a private dependency of kubebuilder?

From my perspective (of course 😸) I'd like kubebuilder to be the semantically versioned umbrella, with each release pinning projects that use it to a specific revision of controller-runtime. Then controller-runtime can remain unversioned or have a different versioning scheme, with the freedom to move quickly knowing that kubebuilder users won't accidentally upgrade.  When users install a new version of kubebuilder, an upgrade procedure can point out or automatically fix incompatibilities.

Anyone using controller-runtime directly is on their own fixing breaking changes, which is probably what they expect from an unversioned library. Maybe a release notes file in the repo could have a list of breaking changes to guide them.

In lieu of the above, what about this until the next major "release" (in the semantic versioning sense):
- Keep the existing `List`, but mark it deprecated, noting that it will be removed at next major release.
- Rename this `List` to `ListVariadic` (or some other, better name). Encourage users to start using it, and note that it will be renamed to `List` at next major release.

That gives users a way to proactively migrate, and a clean upgrade story for those who did (`s/ListVariadic/List/`).

DirectXMan12 commented:
@grantr I we're aiming to set up a version policy (semver) and have a story mostly ironed out for kubebuilder versioning, so we *should* be able to tag first semver release, and then hypothetically make a breaking change, and then do a new major version.

@droot a lot of them that I know do one of a few things:

a) add in alternate names, like suggested by @grantr.   This tends to result in a confusing mess if those are left to languish for too long (or sometimes even if they're not), from what I've seen.
b) do a number of rapid "major" version revs pre-1.0, and then tag 1.0.0 and move to step c
c) "save up" major changes until you hit critical mass for a major version change, then do a large major revision

Of course, we'll want to avoid breaking changes in general, but I think we'll have a short period where we'll want to make one or two and do a few version revs, simply due to the nature of actually getting user feedback (i.e. option b above).

As for `ListVariadic`, since we're still early on, we may just want to do a new major version quickly, after releasing an initial stability version, and avoid confusion from that.  WDYT @droot ?

droot commented:
> How much is controller-runtime a standalone library versus a private dependency of kubebuilder?

Goal is to have controller-runtime a standalone library following proper semver style versioning scheme so that Kubebuilder generated projects are reliably depend on it. As @DirectXMan12 pointed out, we will soon move to the new world to achieve that goal because we were figuring out the dependency story for kubebuilder generated projects. We sort of have a proposal which will work.

> From my perspective (of course 😸) I'd like kubebuilder to be the semantically versioned umbrella, with each release pinning projects that use it to a specific revision of controller-runtime. Then controller-runtime can remain unversioned or have a different versioning scheme, with the freedom to move quickly knowing that kubebuilder users won't accidentally upgrade. When users install a new version of kubebuilder, an upgrade procedure can point out or automatically fix incompatibilities.

Instead of pinning to specific revision (I am assuming you referring to SHA), we can pin the controller-runtime to a minor version (0.1.x say) in Kubebuilder projects so that we don't have to do Kubebuilder release for every patch version on controller-runtime. 

> In lieu of the above, what about this until the next major "release" (in the semantic versioning sense):
Keep the existing List, but mark it deprecated, noting that it will be removed at next major release.
Rename this List to ListVariadic (or some other, better name). Encourage users to start using it, and note that it will be renamed to List at next major release.
That gives users a way to proactively migrate, and a clean upgrade story for those who did (s/ListVariadic/List/).

Sounds reasonable. 

@DirectXMan12 will respond to your post soon.

droot commented:
@grantr we have sorted out the versioning story for controller-runtime/tools and Kubebuilder now uses versioned pkgs of `controller-*`, so we can move forward with these changes now.

DirectXMan12 commented:
looks like it just needs a rebase to fix merge conflicts

grantr commented:
:+1: Hoping to get back to this tomorrow!

grantr commented:
Rebased.

DirectXMan12 commented:
/lgtm
/approve

grantr commented:
There's an issue with the fake client and this new style. 

Currently the fake client's `List` implementation relies on the provided `ListOptions` having a `TypeMeta` in its `Raw` field. See https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/fake/client.go#L81 and [here](https://github.com/knative/eventing/blob/0cb27b7f2540a71bbc82bcf8e4d0a66cf851db71/pkg/controller/eventtype/reconcile.go#L151-L158) for an example of the required workaround in user code.

This PR removes the ability of the user to provide a `Raw` field, and thus breaks the fake client's `List` implementation.

Is there a way to infer the GVK of the resource wrapped by a List object?

DirectXMan12 commented:
/hold
while we figure that out (although the unit test failure should prevent it anyway)

DirectXMan12 commented:
> Is there a way to infer the GVK of the resource wrapped by a List object?

Yeah, you should be able to do similar logic to what we actually do in the client -- the `Scheme` has logic for converting a runtime.Object to a GVK, and then the Kind can be truncated (remove the `List`) to find the underlying Kind.

DirectXMan12 commented:
> This PR removes the ability of the user to provide a Raw field

Huh?  Can you just call `UseListOptions(somethingWithRawSet)`?

grantr commented:
Oh true, that is possible. In practice that would be required for every call to `List` from (well-tested) user code, which kind of defeats the purpose of this change. I'd still like to fix the fake client before merging, or if someone else wants to try that's fine too.

grantr commented:
I updated the fake client to no longer need `Raw.TypeMeta`, updated all the implementations of `List` with the new signature, and updated tests to use the new signature. There's one failing test in certwriter that I haven't figured out yet, but I thought I'd push the new work for review anyway.

grantr commented:
Rebased to squash the fixup commits.

grantr commented on pkg/client/fake/client.go:
```diff
@@ -77,10 +79,23 @@ func (c *fakeClient) Get(ctx context.Context, key client.ObjectKey, obj runtime.
-func (c *fakeClient) List(ctx context.Context, opts *client.ListOptions, list runtime.Object) error {
-	gvk := opts.Raw.TypeMeta.GroupVersionKind()
+func (c *fakeClient) List(ctx context.Context, obj runtime.Object, opts ...client.ListOptionFunc) error {
+	gvk, err := apiutil.GVKForObject(obj, scheme.Scheme)
```
Agreed, though the fake client already restricts itself to `scheme.Scheme` in its constructor https://github.com/kubernetes-sigs/controller-runtime/blob/18a8fbc3f09fd38e210b3f43d4315f42ecd84559/pkg/client/fake/client.go#L50

DirectXMan12 replied:
ah, ok.  Let's do that in a follow-up PR then.  Can you file an issue?

grantr replied:
#137. The workaround for now is to add custom types to `scheme.Scheme` in an init function in the test file. See https://github.com/knative/eventing/blob/e6fb7df8f1b83b6e78f6639f7f7461ee75a9b196/pkg/controller/feed/reconcile_test.go.

DirectXMan12 replied:
thanks!

DirectXMan12 commented:
besides the failing test, this looks good to me.

DirectXMan12 commented:
I don't think so? @mengqiy

grantr commented:
Oh right, Prow isn't used here so `/retest` doesn't work. I don't appear to have permission to try another build - can someone else do that?

Locally I'm consistently getting a different error related to the event recorder:

```
Timeout [1.003 seconds]
recorder
/go/src/sigs.k8s.io/controller-runtime/pkg/internal/recorder/recorder_integration_test.go:36
  recorder
  /go/src/sigs.k8s.io/controller-runtime/pkg/internal/recorder/recorder_integration_test.go:48
    should publish events [It]
    /go/src/sigs.k8s.io/controller-runtime/pkg/internal/recorder/recorder_integration_test.go:49

    Timed out

    /go/src/sigs.k8s.io/controller-runtime/pkg/internal/recorder/recorder_integration_test.go:49
```

DirectXMan12 commented:
@droot @pwittrock we should switch to prow when we get a chance.

DirectXMan12 commented:
I restarted the test.  Let's see if it works, and file a bug about the flakes.  Don't want to have any of those.

DirectXMan12 commented:
/lgtm
/approve

grantr commented:
/hold cancel
(hold was added to block on versioning strategy which is done now)

grantr commented:
Hopefully the final rebase!

grantr commented:
@DirectXMan12 need another lgtm after rebase, please :)

DirectXMan12 commented:
/lgtm
